### PR TITLE
Added epsilon decay

### PR DESCRIPTION
**Pull request by adityauser**
Completed TODO(b/127301657): Decay epsilon based on global step, cf. cl/188907839

### CODE REVIEW

File: tf_agents/agents/dqn/examples/v1/train_eval_gym.py [hunk 1]
```diff
@@ -253,6 +256,9 @@ def train_eval(
 
         global_step_val = global_step_call()
 
+	if global_step_val % epsilon_decay_period == 0 and tf_agent.epsilon_greedy>min_epsilon:
```
Comment (sguada) [inline tf_agents/agents/dqn/examples/v1/train_eval_gym.py]: you need to convert to a TF function so it depends on global_step Variable.

File: tf_agents/agents/dqn/examples/v1/train_eval_gym.py [hunk 2]
```diff
@@ -253,6 +255,9 @@ def train_eval(
...
```
Comment (egonina) [inline tf_agents/agents/dqn/examples/v1/train_eval_gym.py]: add spaces around '>' (tf_agent.epsilon_greedy > min_epsilon)

File: tf_agents/agents/dqn/examples/v1/train_eval_gym.py [hunk 3]
```diff
@@ -253,6 +255,9 @@ def train_eval(
...
+	  tf_agent.epsilon_greedy*=epsilon_decay
```
Comment (egonina) [inline tf_agents/agents/dqn/examples/v1/train_eval_gym.py]: add spaces around '*=', (tf_agent.epsilon_greedy *= epsilon_decay)

**Review by egonina — COMMENTED** [review]
thank you for the PR and apologies for the delay

File: tf_agents/agents/dqn/examples/v1/train_eval_gym.py (hunk 2 above)
Comment (adityauser) [reply tf_agents/agents/dqn/examples/v1/train_eval_gym.py]: Added spaces.

File: tf_agents/agents/dqn/examples/v1/train_eval_gym.py (hunk 3 above)
Comment (adityauser) [reply tf_agents/agents/dqn/examples/v1/train_eval_gym.py]: Added spaces

File: tf_agents/agents/dqn/examples/v1/train_eval_gym.py [hunk 4]
```diff
@@ -253,6 +255,9 @@ def train_eval(
 
         global_step_val = global_step_call()
 
+	if global_step_val % epsilon_decay_period == 0 and tf_agent.epsilon_greedy > min_epsilon:
+	  tf_agent.epsilon_greedy *= epsilon_decay
```
Comment (sguada) [inline tf_agents/agents/dqn/examples/v1/train_eval_gym.py:259]: You cannot do this in TF v1, you need to use a tf.Variable to represent the epsilon_greedy and updated using the global_step.

Comment (adityauser) [reply tf_agents/agents/dqn/examples/v1/train_eval_gym.py:259]: Do you want me to add the following:
Declare epsilon_greedy as tf.Variable
`epsilon_greedy=tf.Variable(1.0, dtype=tf.float64)`

Epsilon decay tf function
```
@tf.function
def decay_epsilon(global_step_val, epsilon_greedy=tf_agent.epsilon_greedy, epsilon_decay_period=epsilon_decay_period, min_epsilon=min_epsilon):
    if global_step_val % epsilon_decay_period == 0 and epsilon_greedy > min_epsilon:
        epsilon_greedy *= epsilon_decay
    return epsilon_greedy
```
Update epsilon
`tf_agent.epsilon_greedy = decay_epsilon(global_step_val)`

**Comment by tfboyd** [comment]
Global step is complicated to handle and handle without creating performance issues.  The last comment was from April 2019.  I am going to close this based on lack of activity.  I am sorry we were not able to bring it to a conclusion; but very much appreciate the effort.  We are focusing on the newer PRs and ensuring we merge or close them in a more timely manner.  Thank you again.

**Pull request closed without merging**